### ISSUE TITLE
Chapter 4 - Early stopping example

### ISSUE THREAD

**Issue by AhmedOmran0** [ageron/handson-ml2#190]
In chapter 4 in the section on early stopping, I found an issue with the way the best model is being retained (I'm using python 3.8.3, scikit-learn 0.23.1)

```
best_epoch = None
best_model = None
for epoch in range(1000):
    sgd_reg.fit(X_train_poly_scaled, y_train)  # continues where it left off
    y_val_predict = sgd_reg.predict(X_val_poly_scaled)
    val_error = mean_squared_error(y_val, y_val_predict)
    if val_error < minimum_val_error:
        minimum_val_error = val_error
        best_epoch = epoch
        best_model = clone(sgd_reg)
```

Here, the best model is declared as a clone of the SGDRegressor for that epoch. However, I found that sklearn.base.clone does not copy the fitted parameters like coef_ and intercept_, but only the regressor parameters. This means that this cloned instance is not useful for making later predictions on new data.

According to the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.base.clone.html) for sklearn.base.clone, this is indeed the case:

> Clone does a deep copy of the model in an estimator without actually copying attached data. 

A simple solution is to do the following instead:

```
from copy import deepcopy
best_model = deepcopy(sgd_reg)
```

**Comment by ageron** [comment]
Hi @AhmedOmran0 ,

Nice catch! Indeed, the `best_model` could not be used directly, it would have to be retrained. That said you generally want to retrain the best model (i.e., the model with the best hyperparameters) on the full dataset, including the training set + the validation set.
Your solution works fine. Another option is to replace `best_model = clone(sgd_reg)` with:

```python
        best_model = sgd_reg
        sgd_reg = clone(sgd_reg)
```

I'll update the book & notebook, thanks! 👍
